### DISCUSSION TITLE
Return ISO datetime string rather than JS Date object

### DISCUSSION THREAD

**Discussion by ericbiewener** [prisma/prisma#5522] [Q&A] [answered]
When sending data back to the client, the date objects returned by a Prisma query have to be converted to strings. Is there away to simply have Prisma not do the string -> date conversion in the first place since this is just throwaway work?

**Comment by nikolasburk** [comment, marked as answer]
Hey @ericbiewener 👋 thanks a lot for raising this. Prisma Client is returning JavaScript `Date` objects [by design](https://www.prisma.io/docs/concepts/components/preview-features/native-types/native-types-mappings#clients-6), unfortunately this can not be configured in the Prisma Client API.

As a potential workaround, you could consider writing a [middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) function that automatically converts `Date` objects to strings so that you don't need to do it manually in your application time every time you're working with a `Date`.

If you'd like to see this natively supported by Prisma, feel free to [open a feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that our Product and Engineering team can look into this 🙌

**Reply by LiamPerson** [reply]
Does anyone have a workaround for this? Really frustrating not having a way to just say to the prisma client to provide dates as integers or strings.

**Comment by 3xau1o** [comment]
There are typescript field type mismatches if types are shared between frontend and backend, UI expects a `Date` still it gets a `string` from js fetch, this can be solved by storing fields as `string` instead of `Date`, at the expense of `@updatedAt` and `@default(now())`

```prisma
model SomeEntity {
 createdAt     DateTime      @default(now())
 updatedAt     DateTime      @updatedAt
}
```

**Reply by JamesLamarRAD** [reply]
@mindlid identified the real problem here. Another example is that when using something like Luxon, you have to pass your date field in as `new Date(myDate)` because the actual property coming back from Prisma is not actually a JavaScript Date. Per the documentation, Prisma Client returns an ISO 8601-formatted string, https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#datetime. So, I think @nikolasburk is somewhat correct here, but the JSON being returned is a string, so the client will never see it as a valid JavaScript Date object. Please help me if I'm missing something here.

So I think the only solution is we have to either convert JSON dates to JavaScript Dates once received or concede our client interfaces to be different than our server ones and make Date properties string.

**Comment by abdessamadely** [comment]
For this I use:

```javascript
const publicArticles = await db.article.findMany()
JSON.parse(JSON.stringify(publicArticles))
```

I don't know with large data if there will be a performance issue, but for now, it's working well for me.

**Reply by abdessamadely** [reply]
I know this one will change the type to `any`, so here's what I do as I still use this solution:

**filename**: types/generic.d.ts
```typescript
export type SerializedStateDates<T> = Omit<
  T,
  'deletedAt' | 'publishedAt' | 'createdAt' | 'updatedAt'
> & {
  publishedAt: string
  createdAt: string
  updatedAt: string
}
```

then I use it as:

```typescript
import { Upload, Serie } from '@prisma/client'
import { SerializedStateDates } from './generic'

type PublicSerieData = SerializedStateDates<Serie> & {
    upload: Upload | null
}
```

**Reply by JohannesSchwegler** [reply]
Great solution! You could even extends the type further to accept the properties as a union.
```ts
export type SerializedStateDates<T, U extends string> = Omit<T, U> & {
  [key in U]: string;
};
```

At then use it like
```ts
SerializedStateDates<Entity, 'createdAt'| 'updatedAt'> 
```

**Reply by rcbevans** [reply]
You can also use generics to replace any Date based property with a string based one:

```
type WithSerializedDates<Type> = {
  [Key in keyof Type]: Type[Key] extends Date
    ? string
    : Type[Key] extends Date | null
    ? string | null
    : Type[Key] extends Date | undefined
    ? string | undefined
    : Type[Key] extends Date | null | undefined
    ? string | null | undefined
    : Type[Key];
};
```

```
/**
 * {
 *  foo: string;
 *  bar: string | null;
 *  baz?: string | undefined;
 * }
 */
type withoutDates = WithSerializedDates<{ foo: Date, bar: Date | null, baz?: Date }>;
```

Unfortunately typescript knows that Date and string aren't overlapping types so just using `as WithSerializedDates<infer prismaResponse>` raises errors.

In my express endpoints I was hoping to be able to `res.json(prismaResponse as WithSerializedDates<infer prismaResponse>)` but typescript raises errors.

> Conversion of type ... may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Types of property 'createdAt' are incompatible.
    Type 'Date' is not comparable to type 'string'.ts(2352)

**Reply by rcbevans** [reply]
For anyone interested, a version of `WithSerializedDates` which works recursively handling child objects or arrays:

```
export type WithSerializedDates<Type> = {
  [Key in keyof Type]: Type[Key] extends Date
    ? string
    : Type[Key] extends Date | null
    ? string | null
    : Type[Key] extends Date | undefined
    ? string | undefined
    : Type[Key] extends Date | null | undefined
    ? string | null | undefined
    : Type[Key] extends Record<PropertyKey, unknown>
    ? WithSerializedDates<Type[Key]>
    : Type[Key] extends Record<PropertyKey, unknown> | null
    ? WithSerializedDates<Type[Key]> | null
    : Type[Key] extends Record<PropertyKey, unknown>[]
    ? WithSerializedDates<Type[Key][number]>[]
    : Type[Key];
};
```

To serialize the dates, I'm cheating and relying on the JSON parser to serialize, but not deserialize back to a date:

```
export const withoutDates = <T>(model: T): WithSerializedDates<T> =>
  JSON.parse(JSON.stringify(model));
```

**Reply by strandchef** [reply]
That is a great idea. I think it can be simplified and even suitable for more use cases like e.g. an optional array of dates with this:

        export type WithSerializedDates<T> = T extends Date
        ? string
        : T extends Array<infer R>
        ? Array<WithSerializedDates<R>>
        : T extends object
        ? { [K in keyof T]: WithSerializedDates<T[K]> }
        : T;

**Comment by psikoi** [comment]
I ran into the same issue with bigints.

My model had a bigint field, but this value was (at most) in the 3-4 billion range, which is above the maximum integer value in postgres, but well below the js number maximum number, so I wanted to store as bigint in the database, but use as `number` in the TS code, and as you said, middlewares don't solve this because of type mismatches.

What I ended up doing is something like this:


## Prisma index file
```typescript
import { PrismaClient, User as PrismaUser } from '@prisma/client'

const prisma = new PrismaClient()

type ModifiedUser = Pick<PrismaUser, "theNameOfTheBigintField"> & { theNameOfTheBigintField: number }

function modifyUser(user: PrismaUser): ModifiedUser {
  // converts bigint to number
  return { ...user,  theNameOfTheBigintField: parseBigInt(user.theNameOfTheBigintField)  }
}

export default prisma;
export { ModifiedUser as User, modifyUser }
```

## Using this model
```typescript
import prisma, { User } from 'pathToPrismaIndex';

const user = await prisma.user
 .findUnique({  where: { id: 1 }}) 
 .then(modifyUser)

  // now, user.theNameOfTheBigintField has the `number` type

```

This is probably not the best way to do this, but it's what I came up with 🤷

**Comment by AndrewSouthpaw** [comment]
I think I'm confused here... in the original response:

> Prisma Client is returning JavaScript Date objects [by design](https://www.prisma.io/docs/concepts/components/preview-features/native-types/native-types-mappings#clients-6), unfortunately this can not be configured in the Prisma Client API.

But the linked documentation says:

> Prisma Client returns all DateTime as ISO 8601-formatted strings.

Did this change? AFAICT Prisma Client is still returning Date objects, which then lead to issues for TS/type checking and communicating values back to the frontend.

**Reply by shuckster** [reply]
Just wanted to add my +1 to this.

The above is exactly what I'm seeing using Prisma with an `sqlite` provider:

1. The generated TypeScript types in `@prisma/client` are `Date`
2. The data returned from a query such as `findMany` is `string` in an ISO-8601 format

So my situation is:

Schema:
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = env("DATABASE_URL")
}

model User {
  id        Int      @id @default(autoincrement())
  name      String   @unique
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

Types:
```ts
import type { User } from '@prisma/client';

/*
User = {
  id: Number;
  name: String;
  createdAt: Date;
  updatedAt: Date;
}
*/
```

App:
```ts
let user = db.user.findUnique({
    where: {
      id: userId,
    },
});

typeof user.createdAt;
// "string"

typeof user.updatedAt;
// "string"
```

Thanks to @abdessamadelhamdany for the info on the helper types to work-around the issue.

@nikolasburk - Any further insight on this please? It took quite a bit of googling to land on this discussion.

**Comment by nklisch** [comment]
I did the following, to both convert the matching types, and convert the Type, inspired by the above post from @rcbevans 
```
type WithSerializedTypes<Type> = {
  [Key in keyof Type]: Type[Key] extends Date
    ? string
    : Type[Key] extends Date | null
    ? string | null
    : Type[Key] extends Date | undefined
    ? string | undefined
    : Type[Key] extends Date | null | undefined
    ? string | null | undefined
    : Type[Key] extends BigInt
    ? number
    : Type[Key] extends BigInt | null
    ? number | null
    : Type[Key] extends BigInt | undefined
    ? number | undefined
    : Type[Key] extends BigInt | null | undefined
    ? number | null | undefined
    : Type[Key];
};

function transformDataTypes<T>(obj: T): WithSerializedTypes<T> {
  for (const [key, value] of Object.entries(obj)) {
    if (typeof value === 'bigint') {
      obj[key] = toNumber(value);
    }
    if (value instanceof Date) {
      obj[key] = value.toISOString();
    }
    if (value === null) {
      delete obj[key];
    }
  }
  return obj as WithSerializedTypes<T>;
}
```

This only does the top level of an object - you'd need to apply it recursively to do a whole object - but for my usecase a flat apply works well.

**Comment by mubaidr** [comment]
I think, the type here is `date` as well. But over the channel (through API) these `date`s gets serialized to `string`. This leads to manually updating `Prisma` to expect string instead of dates. 
An option to specify `string` as default type for `date` columns would be great!

**Comment by SSabev** [comment]
Terrible answer, but we just did z.any() in our output schema.

**Comment by tzezar** [comment]
Has any work been done in this direction?

**Comment by jove4015** [comment]
Just wanted to add some more use cases here - another issue we're facing is the case of db.Date objects, where in the database what's really stored is just an ISO date (ie, "2024-08-01"), but Prisma then converts this to a date object by adding essentially adding "T00:00:00Z" and then converting to a Date object with a time component.  Ultimately, this creates time zone problems where the iso date winds up being converted and then converted back to the day before (or potentially the next day, although it usually works the other way).  It would be much, much better for us if time was never associated with this value and it always stayed statically just the Date.  Having to convert back and forth with this all the time is ... trying at its best.  We will never see the end of time zone problems because of this issue and just plain strings with no Date conversion would be tremendously helpful.

**Comment by CyanoFresh** [comment]
that's disgusting.
even with plain pg driver we can override this behaviour: https://github.com/brianc/node-postgres/issues/1844#issuecomment-862436757
even using prisma.queryRawUnsafe converts Dates (without time) to js Date with time

**Comment by JonRC** [comment]
I created a util to convert all Dates in an object to string deeply.

```TS
export type ClientOf<T> = {
  [K in keyof T]: NonNullable<T[K]> extends Date
    ? string | Exclude<T[K], Date>
    : T[K] extends object
      ? ClientOf<T[K]>
      : T[K];
};
```
Example  
```TS
type User = {
  name: string
  updatedAt: Date | null
}

type UserClient = ClientOf<User>

// UserClient will be {name: string, updatedAt: string | null}
```

**Comment by YanceyOfficial** [comment]
Yeah, Prisma always returns the Date object type, but Protobuf can only define string or number types. I don't know whether there is an automatic way to deal with it.

**Comment by christian-gama** [comment]
I was able to solve it using Prisma Extensions.

```typescript
const prisma = new PrismaClient();

const extension = prisma.$extends({
	result: {
		$allModels: {
			createdAt: {
				needs: { createdAt: true } as any,
				compute(data: { createdAt: Date }) {
					return DateTime.fromJSDate(data.createdAt).toISO();
				},
			},
			updatedAt: {
				needs: { updatedAt: true } as any,
				compute(data: { updatedAt: Date }) {
					return DateTime.fromJSDate(data.updatedAt).toISO();
				},
			},
		},
	},
});

const user = await extension.user.findFirst()
console.log(user.createdAt) // Type is automatically `string` and the value is in ISO format
```

**Comment by inmanenz** [comment]
I haven't found any recent issues related to this, so I opened #25827. Drizzle has a pretty great solution, hopefully Prisma can add something similar!
